File: w03/P3.py
class Event:
  def __init__(self, arrival, duration):
    self.arrival = arrival
    self.duration = duration

class Queue:
  def __init__(self):
    self.events = []

  def size(self):
    return len(self.events)

  def enQueue(self, evt):
    self.events.append(evt)

  def deQueue(self):
    self.events.pop(0)

  def peek(self):
    return self.events[0]

  def isEmpty(self):
    return len(self.events) == 0

eventQ = Queue()
eventQ.enQueue(Event(0, 5))
eventQ.enQueue(Event(3, 3))
eventQ.enQueue(Event(4, 4))
# eventQ.enQueue(Event(100, 4))

nextAvailableTime = 0
totalWaiting = 0
maxWaiting = 0
size = eventQ.size()

while not eventQ.isEmpty():
  evt = eventQ.peek()
  eventQ.deQueue()
  nextAvailableTime = max(nextAvailableTime, evt.arrival)
  totalWaiting += (nextAvailableTime - evt.arrival)
  maxWaiting = max(maxWaiting, nextAvailableTime - evt.arrival)
  nextAvailableTime = nextAvailableTime + evt.duration

print("Total waiting time", totalWaiting, ", average waiting time", totalWaiting / size, ", and maximum waiting time", maxWaiting)

File: w03/P3.java
package w03;

public class P3 {
  public static void main(String[] args) {
    QueueApplication.eventSimulation();
  }  
}

class QueueApplication {
  static class Event {
    int arrival;
    int duration;

    public Event(int a, int d) {
      arrival = a;
      duration = d;
    }
  }

  static void eventSimulation() {
    ArrayQueue<Event> events = new ArrayQueue<Event>();
    events.enQueue(new Event(0, 5));
    events.enQueue(new Event(3, 3));
    events.enQueue(new Event(4, 4));
    events.enQueue(new Event(100, 4));

    int n = events.size();

    int nextAvailableTime = 0;
    int totalWaiting = 0;

    while (!events.isEmpty()) {
      Event evt = events.peekFront();
      events.deQueue();
      nextAvailableTime = Math.max(nextAvailableTime, evt.arrival);
      totalWaiting += (nextAvailableTime - evt.arrival);
      nextAvailableTime = nextAvailableTime + evt.duration;
    }

    System.out.printf("Total waiting time %d and average waiting time %.2f\n", totalWaiting, 1.0 * totalWaiting / n);
  }
}
